File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/CommodityPhysicalEuropeanExercise.cpp
// CommodityPhysicalEuropeanExercise.cpp 
#include "CommodityPhysicalEuropeanExercise.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CommodityPhysicalEuropeanExercise::CommodityPhysicalEuropeanExercise(TiXmlNode* xmlNode)
: Exercise(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //expirationDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationDateNode = xmlNode->FirstChildElement("expirationDate");

   if(expirationDateNode){expirationDateIsNull_ = false;}
   else{expirationDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationDateNode , address : " << expirationDateNode << std::endl;
   #endif
   if(expirationDateNode)
   {
      if(expirationDateNode->Attribute("href") || expirationDateNode->Attribute("id"))
      {
          if(expirationDateNode->Attribute("id"))
          {
             expirationDateIDRef_ = expirationDateNode->Attribute("id");
             expirationDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(expirationDateNode));
             expirationDate_->setID(expirationDateIDRef_);
             IDManager::instance().setID(expirationDateIDRef_,expirationDate_);
          }
          else if(expirationDateNode->Attribute("href")) { expirationDateIDRef_ = expirationDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(expirationDateNode));}
   }

   //expirationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationDatesNode = xmlNode->FirstChildElement("expirationDates");

   if(expirationDatesNode){expirationDatesIsNull_ = false;}
   else{expirationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationDatesNode , address : " << expirationDatesNode << std::endl;
   #endif
   if(expirationDatesNode)
   {
      if(expirationDatesNode->Attribute("href") || expirationDatesNode->Attribute("id"))
      {
          if(expirationDatesNode->Attribute("id"))
          {
             expirationDatesIDRef_ = expirationDatesNode->Attribute("id");
             expirationDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>(new AdjustableRelativeOrPeriodicDates2(expirationDatesNode));
             expirationDates_->setID(expirationDatesIDRef_);
             IDManager::instance().setID(expirationDatesIDRef_,expirationDates_);
          }
          else if(expirationDatesNode->Attribute("href")) { expirationDatesIDRef_ = expirationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>(new AdjustableRelativeOrPeriodicDates2(expirationDatesNode));}
   }

   //relativeExpirationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relativeExpirationDatesNode = xmlNode->FirstChildElement("relativeExpirationDates");

   if(relativeExpirationDatesNode){relativeExpirationDatesIsNull_ = false;}
   else{relativeExpirationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relativeExpirationDatesNode , address : " << relativeExpirationDatesNode << std::endl;
   #endif
   if(relativeExpirationDatesNode)
   {
      if(relativeExpirationDatesNode->Attribute("href") || relativeExpirationDatesNode->Attribute("id"))
      {
          if(relativeExpirationDatesNode->Attribute("id"))
          {
             relativeExpirationDatesIDRef_ = relativeExpirationDatesNode->Attribute("id");
             relativeExpirationDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeExpirationDatesNode));
             relativeExpirationDates_->setID(relativeExpirationDatesIDRef_);
             IDManager::instance().setID(relativeExpirationDatesIDRef_,relativeExpirationDates_);
          }
          else if(relativeExpirationDatesNode->Attribute("href")) { relativeExpirationDatesIDRef_ = relativeExpirationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relativeExpirationDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeExpirationDatesNode));}
   }

   //expirationTimeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationTimeNode = xmlNode->FirstChildElement("expirationTime");

   if(expirationTimeNode){expirationTimeIsNull_ = false;}
   else{expirationTimeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationTimeNode , address : " << expirationTimeNode << std::endl;
   #endif
   if(expirationTimeNode)
   {
      if(expirationTimeNode->Attribute("href") || expirationTimeNode->Attribute("id"))
      {
          if(expirationTimeNode->Attribute("id"))
          {
             expirationTimeIDRef_ = expirationTimeNode->Attribute("id");
             expirationTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(expirationTimeNode));
             expirationTime_->setID(expirationTimeIDRef_);
             IDManager::instance().setID(expirationTimeIDRef_,expirationTime_);
          }
          else if(expirationTimeNode->Attribute("href")) { expirationTimeIDRef_ = expirationTimeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(expirationTimeNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableOrRelativeDate> CommodityPhysicalEuropeanExercise::getExpirationDate()
{
   if(!this->expirationDateIsNull_){
        if(expirationDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(expirationDateIDRef_));
        }else{
             return this->expirationDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<AdjustableRelativeOrPeriodicDates2> CommodityPhysicalEuropeanExercise::getExpirationDates()
{
   if(!this->expirationDatesIsNull_){
        if(expirationDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableRelativeOrPeriodicDates2>(IDManager::instance().getID(expirationDatesIDRef_));
        }else{
             return this->expirationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>();
   }
}
boost::shared_ptr<CommodityRelativeExpirationDates> CommodityPhysicalEuropeanExercise::getRelativeExpirationDates()
{
   if(!this->relativeExpirationDatesIsNull_){
        if(relativeExpirationDatesIDRef_ != ""){
             return boost::shared_static_cast<CommodityRelativeExpirationDates>(IDManager::instance().getID(relativeExpirationDatesIDRef_));
        }else{
             return this->relativeExpirationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CommodityRelativeExpirationDates>();
   }
}
boost::shared_ptr<PrevailingTime> CommodityPhysicalEuropeanExercise::getExpirationTime()
{
   if(!this->expirationTimeIsNull_){
        if(expirationTimeIDRef_ != ""){
             return boost::shared_static_cast<PrevailingTime>(IDManager::instance().getID(expirationTimeIDRef_));
        }else{
             return this->expirationTime_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PrevailingTime>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/CommodityPhysicalEuropeanExercise.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CommodityPhysicalEuropeanExercise : Exercise
    {
        public CommodityPhysicalEuropeanExercise(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode expirationDateNode = xmlNode.SelectSingleNode("expirationDate");
            
            if (expirationDateNode != null)
            {
                if (expirationDateNode.Attributes["href"] != null || expirationDateNode.Attributes["id"] != null) 
                {
                    if (expirationDateNode.Attributes["id"] != null) 
                    {
                        expirationDateIDRef_ = expirationDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(expirationDateNode);
                        IDManager.SetID(expirationDateIDRef_, ob);
                    }
                    else if (expirationDateNode.Attributes["href"] != null)
                    {
                        expirationDateIDRef_ = expirationDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationDate_ = new AdjustableOrRelativeDate(expirationDateNode);
                    }
                }
                else
                {
                    expirationDate_ = new AdjustableOrRelativeDate(expirationDateNode);
                }
            }
            
        
            XmlNode expirationDatesNode = xmlNode.SelectSingleNode("expirationDates");
            
            if (expirationDatesNode != null)
            {
                if (expirationDatesNode.Attributes["href"] != null || expirationDatesNode.Attributes["id"] != null) 
                {
                    if (expirationDatesNode.Attributes["id"] != null) 
                    {
                        expirationDatesIDRef_ = expirationDatesNode.Attributes["id"].Value;
                        AdjustableRelativeOrPeriodicDates2 ob = new AdjustableRelativeOrPeriodicDates2(expirationDatesNode);
                        IDManager.SetID(expirationDatesIDRef_, ob);
                    }
                    else if (expirationDatesNode.Attributes["href"] != null)
                    {
                        expirationDatesIDRef_ = expirationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationDates_ = new AdjustableRelativeOrPeriodicDates2(expirationDatesNode);
                    }
                }
                else
                {
                    expirationDates_ = new AdjustableRelativeOrPeriodicDates2(expirationDatesNode);
                }
            }
            
        
            XmlNode relativeExpirationDatesNode = xmlNode.SelectSingleNode("relativeExpirationDates");
            
            if (relativeExpirationDatesNode != null)
            {
                if (relativeExpirationDatesNode.Attributes["href"] != null || relativeExpirationDatesNode.Attributes["id"] != null) 
                {
                    if (relativeExpirationDatesNode.Attributes["id"] != null) 
                    {
                        relativeExpirationDatesIDRef_ = relativeExpirationDatesNode.Attributes["id"].Value;
                        CommodityRelativeExpirationDates ob = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                        IDManager.SetID(relativeExpirationDatesIDRef_, ob);
                    }
                    else if (relativeExpirationDatesNode.Attributes["href"] != null)
                    {
                        relativeExpirationDatesIDRef_ = relativeExpirationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relativeExpirationDates_ = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                    }
                }
                else
                {
                    relativeExpirationDates_ = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                }
            }
            
        
            XmlNode expirationTimeNode = xmlNode.SelectSingleNode("expirationTime");
            
            if (expirationTimeNode != null)
            {
                if (expirationTimeNode.Attributes["href"] != null || expirationTimeNode.Attributes["id"] != null) 
                {
                    if (expirationTimeNode.Attributes["id"] != null) 
                    {
                        expirationTimeIDRef_ = expirationTimeNode.Attributes["id"].Value;
                        PrevailingTime ob = new PrevailingTime(expirationTimeNode);
                        IDManager.SetID(expirationTimeIDRef_, ob);
                    }
                    else if (expirationTimeNode.Attributes["href"] != null)
                    {
                        expirationTimeIDRef_ = expirationTimeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationTime_ = new PrevailingTime(expirationTimeNode);
                    }
                }
                else
                {
                    expirationTime_ = new PrevailingTime(expirationTimeNode);
                }
            }
            
        
        }
        
    
        #region ExpirationDate_
        private AdjustableOrRelativeDate expirationDate_;
        public AdjustableOrRelativeDate ExpirationDate_
        {
            get
            {
                if (this.expirationDate_ != null)
                {
                    return this.expirationDate_; 
                }
                else if (this.expirationDateIDRef_ != null)
                {
                    expirationDate_ = IDManager.getID(expirationDateIDRef_) as AdjustableOrRelativeDate;
                    return this.expirationDate_; 
                }
                else
                {
                      return this.expirationDate_; 
                }
            }
            set
            {
                if (this.expirationDate_ != value)
                {
                    this.expirationDate_ = value;
                }
            }
        }
        #endregion
        
        public string expirationDateIDRef_ { get; set; }
        #region ExpirationDates_
        private AdjustableRelativeOrPeriodicDates2 expirationDates_;
        public AdjustableRelativeOrPeriodicDates2 ExpirationDates_
        {
            get
            {
                if (this.expirationDates_ != null)
                {
                    return this.expirationDates_; 
                }
                else if (this.expirationDatesIDRef_ != null)
                {
                    expirationDates_ = IDManager.getID(expirationDatesIDRef_) as AdjustableRelativeOrPeriodicDates2;
                    return this.expirationDates_; 
                }
                else
                {
                      return this.expirationDates_; 
                }
            }
            set
            {
                if (this.expirationDates_ != value)
                {
                    this.expirationDates_ = value;
                }
            }
        }
        #endregion
        
        public string expirationDatesIDRef_ { get; set; }
        #region RelativeExpirationDates_
        private CommodityRelativeExpirationDates relativeExpirationDates_;
        public CommodityRelativeExpirationDates RelativeExpirationDates_
        {
            get
            {
                if (this.relativeExpirationDates_ != null)
                {
                    return this.relativeExpirationDates_; 
                }
                else if (this.relativeExpirationDatesIDRef_ != null)
                {
                    relativeExpirationDates_ = IDManager.getID(relativeExpirationDatesIDRef_) as CommodityRelativeExpirationDates;
                    return this.relativeExpirationDates_; 
                }
                else
                {
                      return this.relativeExpirationDates_; 
                }
            }
            set
            {
                if (this.relativeExpirationDates_ != value)
                {
                    this.relativeExpirationDates_ = value;
                }
            }
        }
        #endregion
        
        public string relativeExpirationDatesIDRef_ { get; set; }
        #region ExpirationTime_
        private PrevailingTime expirationTime_;
        public PrevailingTime ExpirationTime_
        {
            get
            {
                if (this.expirationTime_ != null)
                {
                    return this.expirationTime_; 
                }
                else if (this.expirationTimeIDRef_ != null)
                {
                    expirationTime_ = IDManager.getID(expirationTimeIDRef_) as PrevailingTime;
                    return this.expirationTime_; 
                }
                else
                {
                      return this.expirationTime_; 
                }
            }
            set
            {
                if (this.expirationTime_ != value)
                {
                    this.expirationTime_ = value;
                }
            }
        }
        #endregion
        
        public string expirationTimeIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

